File: ios/runanywhere-sdks-latest/sdk/runanywhere-swift/Sources/RunAnywhere/Public/Extensions/LLM/RunAnywhere+TextGeneration.swift
//
//  RunAnywhere+TextGeneration.swift
//  RunAnywhere SDK
//
//  Public API for text generation (LLM) operations.
//  Calls C++ directly via CppBridge.LLM for all operations.
//  Events are emitted by C++ layer via CppEventBridge.
//

import CRACommons
import Foundation

// MARK: - Text Generation

public extension RunAnywhere {

    /// Simple text generation with automatic event publishing
    /// - Parameter prompt: The text prompt
    /// - Returns: Generated response (text only)
    static func chat(_ prompt: String) async throws -> String {
        let result = try await generate(prompt, options: nil)
        return result.text
    }

    /// Generate text with full metrics and analytics
    /// - Parameters:
    ///   - prompt: The text prompt
    ///   - options: Generation options (optional)
    /// - Returns: GenerationResult with full metrics including thinking tokens, timing, performance, etc.
    /// - Note: Events are automatically dispatched via C++ layer
    static func generate(
        _ prompt: String,
        options: LLMGenerationOptions? = nil
    ) async throws -> LLMGenerationResult {
        guard isInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        try await ensureServicesReady()

        // Get handle from CppBridge.LLM
        let handle = try await CppBridge.LLM.shared.getHandle()

        // Verify model is loaded
        guard await CppBridge.LLM.shared.isLoaded else {
            throw SDKError.llm(.notInitialized, "LLM model not loaded")
        }

        let modelId = await CppBridge.LLM.shared.currentModelId ?? "unknown"
        let opts = options ?? LLMGenerationOptions()

        let startTime = Date()

        // Build C options
        var cOptions = rac_llm_options_t()
        cOptions.max_tokens = Int32(opts.maxTokens)
        cOptions.temperature = opts.temperature
        cOptions.top_p = opts.topP
        cOptions.streaming_enabled = RAC_FALSE

        SDKLogger.llm.info("[PARAMS] generate: temperature=\(cOptions.temperature), top_p=\(cOptions.top_p), max_tokens=\(cOptions.max_tokens), system_prompt=\(opts.systemPrompt != nil ? "set(\(opts.systemPrompt!.count) chars)" : "nil"), streaming=\(cOptions.streaming_enabled == RAC_TRUE)")

        // Generate (C++ emits events) - wrap in system_prompt lifetime scope
        var llmResult = rac_llm_result_t()
        let generateResult: rac_result_t
        if let systemPrompt = opts.systemPrompt {
            generateResult = systemPrompt.withCString { sysPromptPtr in
                cOptions.system_prompt = sysPromptPtr
                return prompt.withCString { promptPtr in
                    rac_llm_component_generate(handle, promptPtr, &cOptions, &llmResult)
                }
            }
        } else {
            cOptions.system_prompt = nil
            generateResult = prompt.withCString { promptPtr in
                rac_llm_component_generate(handle, promptPtr, &cOptions, &llmResult)
            }
        }

        guard generateResult == RAC_SUCCESS else {
            throw SDKError.llm(.generationFailed, "Generation failed: \(generateResult)")
        }

        let endTime = Date()
        let totalTimeMs = endTime.timeIntervalSince(startTime) * 1000

        // Extract result
        let generatedText: String
        if let textPtr = llmResult.text {
            generatedText = String(cString: textPtr)
        } else {
            generatedText = ""
        }
        let inputTokens = Int(llmResult.prompt_tokens)
        let outputTokens = Int(llmResult.completion_tokens)
        let tokensPerSecond = llmResult.tokens_per_second > 0 ? Double(llmResult.tokens_per_second) : 0

        return LLMGenerationResult(
            text: generatedText,
            thinkingContent: nil,
            inputTokens: inputTokens,
            tokensUsed: outputTokens,
            modelUsed: modelId,
            latencyMs: totalTimeMs,
            framework: "llamacpp",
            tokensPerSecond: tokensPerSecond,
            timeToFirstTokenMs: nil,
            thinkingTokens: 0,
            responseTokens: outputTokens
        )
    }

    /// Streaming text generation with complete analytics
    ///
    /// Returns both a token stream for real-time display and a task that resolves to complete metrics.
    ///
    /// Example usage:
    /// ```swift
    /// let result = try await RunAnywhere.generateStream(prompt)
    ///
    /// // Display tokens in real-time
    /// for try await token in result.stream {
    ///     print(token, terminator: "")
    /// }
    ///
    /// // Get complete analytics after streaming finishes
    /// let metrics = try await result.result.value
    /// print("Speed: \(metrics.performanceMetrics.tokensPerSecond) tok/s")
    /// print("Tokens: \(metrics.tokensUsed)")
    /// print("Time: \(metrics.latencyMs)ms")
    /// ```
    ///
    /// - Parameters:
    ///   - prompt: The text prompt
    ///   - options: Generation options (optional)
    /// - Returns: StreamingResult containing both the token stream and final metrics task
    static func generateStream(
        _ prompt: String,
        options: LLMGenerationOptions? = nil
    ) async throws -> LLMStreamingResult {
        guard isInitialized else {
            throw SDKError.general(.notInitialized, "SDK not initialized")
        }

        try await ensureServicesReady()

        let handle = try await CppBridge.LLM.shared.getHandle()

        guard await CppBridge.LLM.shared.isLoaded else {
            throw SDKError.llm(.notInitialized, "LLM model not loaded")
        }

        let modelId = await CppBridge.LLM.shared.currentModelId ?? "unknown"
        let opts = options ?? LLMGenerationOptions()

        let collector = LLMStreamingMetricsCollector(modelId: modelId, promptLength: prompt.count)

        var cOptions = rac_llm_options_t()
        cOptions.max_tokens = Int32(opts.maxTokens)
        cOptions.temperature = opts.temperature
        cOptions.top_p = opts.topP
        cOptions.streaming_enabled = RAC_TRUE

        SDKLogger.llm.info("[PARAMS] generateStream: temperature=\(cOptions.temperature), top_p=\(cOptions.top_p), max_tokens=\(cOptions.max_tokens), system_prompt=\(opts.systemPrompt != nil ? "set(\(opts.systemPrompt!.count) chars)" : "nil"), streaming=\(cOptions.streaming_enabled == RAC_TRUE)")

        let stream = createTokenStream(
            prompt: prompt,
            handle: handle,
            options: cOptions,
            collector: collector,
            systemPrompt: opts.systemPrompt
        )

        let resultTask = Task<LLMGenerationResult, Error> {
            try await collector.waitForResult()
        }

        return LLMStreamingResult(stream: stream, result: resultTask)
    }

    // MARK: - Private Streaming Helpers

    private static func createTokenStream(
        prompt: String,
        handle: UnsafeMutableRawPointer,
        options: rac_llm_options_t,
        collector: LLMStreamingMetricsCollector,
        systemPrompt: String? = nil
    ) -> AsyncThrowingStream<String, Error> {
        AsyncThrowingStream<String, Error> { continuation in
            Task {
                do {
                    await collector.markStart()

                    let context = LLMStreamCallbackContext(continuation: continuation, collector: collector)
                    let contextPtr = Unmanaged.passRetained(context).toOpaque()

                    let callbacks = LLMStreamCallbacks.create()
                    var cOptions = options

                    let callCFunction: () -> rac_result_t = {
                        prompt.withCString { promptPtr in
                            rac_llm_component_generate_stream(
                                handle,
                                promptPtr,
                                &cOptions,
                                callbacks.token,
                                callbacks.complete,
                                callbacks.error,
                                contextPtr
                            )
                        }
                    }

                    let streamResult: rac_result_t
                    if let systemPrompt = systemPrompt {
                        streamResult = systemPrompt.withCString { sysPtr in
                            cOptions.system_prompt = sysPtr
                            return callCFunction()
                        }
                    } else {
                        cOptions.system_prompt = nil
                        streamResult = callCFunction()
                    }

                    if streamResult != RAC_SUCCESS {
                        Unmanaged<LLMStreamCallbackContext>.fromOpaque(contextPtr).release()
                        let error = SDKError.llm(.generationFailed, "Stream generation failed: \(streamResult)")
                        continuation.finish(throwing: error)
                        await collector.markFailed(error)
                    }
                } catch {
                    continuation.finish(throwing: error)
                    await collector.markFailed(error)
                }
            }
        }
    }

}

// MARK: - Streaming Callbacks

private enum LLMStreamCallbacks {
    typealias TokenFn = rac_llm_component_token_callback_fn
    typealias CompleteFn = rac_llm_component_complete_callback_fn
    typealias ErrorFn = rac_llm_component_error_callback_fn

    struct Callbacks {
        let token: TokenFn
        let complete: CompleteFn
        let error: ErrorFn
    }

    static func create() -> Callbacks {
        let tokenCallback: TokenFn = { tokenPtr, userData -> rac_bool_t in
            guard let tokenPtr = tokenPtr, let userData = userData else { return RAC_TRUE }
            let ctx = Unmanaged<LLMStreamCallbackContext>.fromOpaque(userData).takeUnretainedValue()
            let token = String(cString: tokenPtr)
            Task {
                await ctx.collector.recordToken(token)
                ctx.continuation.yield(token)
            }
            return RAC_TRUE
        }

        let completeCallback: CompleteFn = { _, userData in
            guard let userData = userData else { return }
            let ctx = Unmanaged<LLMStreamCallbackContext>.fromOpaque(userData).takeUnretainedValue()
            ctx.continuation.finish()
            Task { await ctx.collector.markComplete() }
        }

        let errorCallback: ErrorFn = { _, errorMsg, userData in
            guard let userData = userData else { return }
            let ctx = Unmanaged<LLMStreamCallbackContext>.fromOpaque(userData).takeUnretainedValue()
            let message = errorMsg.map { String(cString: $0) } ?? "Unknown error"
            let error = SDKError.llm(.generationFailed, message)
            ctx.continuation.finish(throwing: error)
            Task { await ctx.collector.markFailed(error) }
        }

        return Callbacks(token: tokenCallback, complete: completeCallback, error: errorCallback)
    }
}

// MARK: - Streaming Callback Context

private final class LLMStreamCallbackContext: @unchecked Sendable {
    let continuation: AsyncThrowingStream<String, Error>.Continuation
    let collector: LLMStreamingMetricsCollector

    init(continuation: AsyncThrowingStream<String, Error>.Continuation, collector: LLMStreamingMetricsCollector) {
        self.continuation = continuation
        self.collector = collector
    }
}

// MARK: - Streaming Metrics Collector

/// Internal actor for collecting streaming metrics
private actor LLMStreamingMetricsCollector {
    private let modelId: String
    private let promptLength: Int

    private var startTime: Date?
    private var firstTokenTime: Date?
    private var fullText = ""
    private var tokenCount = 0
    private var firstTokenRecorded = false
    private var isComplete = false
    private var error: Error?
    private var resultContinuation: CheckedContinuation<LLMGenerationResult, Error>?

    init(modelId: String, promptLength: Int) {
        self.modelId = modelId
        self.promptLength = promptLength
    }

    func markStart() {
        startTime = Date()
    }

    func recordToken(_ token: String) {
        fullText += token
        tokenCount += 1

        if !firstTokenRecorded {
            firstTokenRecorded = true
            firstTokenTime = Date()
        }
    }

    func markComplete() {
        isComplete = true
        if let continuation = resultContinuation {
            continuation.resume(returning: buildResult())
            resultContinuation = nil
        }
    }

    func markFailed(_ error: Error) {
        self.error = error
        if let continuation = resultContinuation {
            continuation.resume(throwing: error)
            resultContinuation = nil
        }
    }

    func waitForResult() async throws -> LLMGenerationResult {
        if isComplete {
            return buildResult()
        }
        if let error = error {
            throw error
        }
        return try await withCheckedThrowingContinuation { continuation in
            resultContinuation = continuation
        }
    }

    private func buildResult() -> LLMGenerationResult {
        let endTime = Date()
        let latencyMs = (startTime.map { endTime.timeIntervalSince($0) } ?? 0) * 1000

        var timeToFirstTokenMs: Double?
        if let start = startTime, let firstToken = firstTokenTime {
            timeToFirstTokenMs = firstToken.timeIntervalSince(start) * 1000
        }

        // Use actual token count from streaming callbacks, not character estimation (fixes #339)
        let outputTokens = max(1, tokenCount)
        let totalTimeSec = latencyMs / 1000.0
        let tokensPerSecond = totalTimeSec > 0 ? Double(outputTokens) / totalTimeSec : 0

        return LLMGenerationResult(
            text: fullText,
            thinkingContent: nil,
            inputTokens: 0,
            tokensUsed: outputTokens,
            modelUsed: modelId,
            latencyMs: latencyMs,
            framework: "llamacpp",
            tokensPerSecond: tokensPerSecond,
            timeToFirstTokenMs: timeToFirstTokenMs,
            thinkingTokens: 0,
            responseTokens: outputTokens
        )
    }
}

File: ios/runanywhere-sdks-latest/sdk/runanywhere-react-native/packages/core/src/Public/Extensions/RunAnywhere+TextGeneration.ts
/**
 * RunAnywhere+TextGeneration.ts
 *
 * Text generation (LLM) extension for RunAnywhere SDK.
 * Uses backend-agnostic rac_llm_component_* C++ APIs via the core native module.
 * The actual backend (LlamaCPP, etc.) must be registered by installing
 * and importing the appropriate backend package (e.g., @runanywhere/llamacpp).
 *
 * Matches iOS: RunAnywhere+TextGeneration.swift
 */

import { EventBus } from '../Events';
import {
  requireNativeModule,
  isNativeModuleAvailable,
} from '../../native';
import type { GenerationOptions, GenerationResult } from '../../types';
import { ExecutionTarget, HardwareAcceleration } from '../../types';
import { SDKLogger } from '../../Foundation/Logging/Logger/SDKLogger';
import type {
  LLMStreamingResult,
  LLMGenerationResult,
} from '../../types/LLMTypes';

const logger = new SDKLogger('RunAnywhere.TextGeneration');

// ============================================================================
// Text Generation (LLM) Extension - Backend Agnostic
// ============================================================================

/**
 * Load an LLM model by ID or path
 *
 * Matches iOS: `RunAnywhere.loadModel(_:)`
 * @throws Error if no LLM backend is registered
 */
export async function loadModel(
  modelPathOrId: string,
  config?: Record<string, unknown>
): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    logger.warning('Native module not available for loadModel');
    return false;
  }
  const native = requireNativeModule();
  return native.loadTextModel(
    modelPathOrId,
    config ? JSON.stringify(config) : undefined
  );
}

/**
 * Check if an LLM model is loaded
 * Matches iOS: `RunAnywhere.isModelLoaded`
 */
export async function isModelLoaded(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule();
  return native.isTextModelLoaded();
}

/**
 * Unload the currently loaded LLM model
 * Matches iOS: `RunAnywhere.unloadModel()`
 */
export async function unloadModel(): Promise<boolean> {
  if (!isNativeModuleAvailable()) {
    return false;
  }
  const native = requireNativeModule();
  return native.unloadTextModel();
}

/**
 * Simple chat - returns just the text response
 * Matches Swift SDK: RunAnywhere.chat(_:)
 */
export async function chat(prompt: string): Promise<string> {
  const result = await generate(prompt);
  return result.text;
}

/**
 * Text generation with options and full metrics
 * Matches Swift SDK: RunAnywhere.generate(_:options:)
 */
export async function generate(
  prompt: string,
  options?: GenerationOptions
): Promise<GenerationResult> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }
  const native = requireNativeModule();

  const optionsJson = JSON.stringify({
    max_tokens: options?.maxTokens ?? 1000,
    temperature: options?.temperature ?? 0.7,
    system_prompt: options?.systemPrompt ?? null,
  });

  const resultJson = await native.generate(prompt, optionsJson);

  try {
    const result = JSON.parse(resultJson);
    return {
      text: result.text ?? '',
      thinkingContent: result.thinkingContent,
      tokensUsed: result.tokensUsed ?? 0,
      modelUsed: result.modelUsed ?? 'unknown',
      latencyMs: result.latencyMs ?? 0,
      executionTarget: result.executionTarget ?? 0,
      savedAmount: result.savedAmount ?? 0,
      framework: result.framework,
      hardwareUsed: result.hardwareUsed ?? 0,
      memoryUsed: result.memoryUsed ?? 0,
      performanceMetrics: {
        timeToFirstTokenMs: result.performanceMetrics?.timeToFirstTokenMs,
        tokensPerSecond: result.performanceMetrics?.tokensPerSecond,
        inferenceTimeMs:
          result.performanceMetrics?.inferenceTimeMs ?? result.latencyMs ?? 0,
      },
      thinkingTokens: result.thinkingTokens,
      responseTokens: result.responseTokens ?? result.tokensUsed ?? 0,
    };
  } catch {
    if (resultJson.includes('error')) {
      throw new Error(resultJson);
    }
    return {
      text: resultJson,
      tokensUsed: 0,
      modelUsed: 'unknown',
      latencyMs: 0,
      executionTarget: ExecutionTarget.OnDevice,
      savedAmount: 0,
      hardwareUsed: HardwareAcceleration.CPU,
      memoryUsed: 0,
      performanceMetrics: {
        inferenceTimeMs: 0,
      },
      responseTokens: 0,
    };
  }
}

/**
 * Streaming text generation with async iterator
 *
 * Returns a LLMStreamingResult containing:
 * - stream: AsyncIterable<string> for consuming tokens
 * - result: Promise<LLMGenerationResult> for final metrics
 * - cancel: Function to cancel generation
 *
 * Matches Swift SDK: RunAnywhere.generateStream(_:options:)
 *
 * Example usage:
 * ```typescript
 * const streaming = await generateStream(prompt);
 *
 * // Display tokens in real-time
 * for await (const token of streaming.stream) {
 *   console.log(token);
 * }
 *
 * // Get complete analytics after streaming finishes
 * const metrics = await streaming.result;
 * console.log(`Speed: ${metrics.tokensPerSecond} tok/s`);
 * ```
 */
export async function generateStream(
  prompt: string,
  options?: GenerationOptions
): Promise<LLMStreamingResult> {
  if (!isNativeModuleAvailable()) {
    throw new Error('Native module not available');
  }

  const native = requireNativeModule();
  const startTime = Date.now();
  let firstTokenTime: number | null = null;
  let cancelled = false;
  let fullText = '';
  let tokenCount = 0;
  let resolveResult: ((result: LLMGenerationResult) => void) | null = null;
  let rejectResult: ((error: Error) => void) | null = null;

  const optionsJson = JSON.stringify({
    max_tokens: options?.maxTokens ?? 1000,
    temperature: options?.temperature ?? 0.7,
    system_prompt: options?.systemPrompt ?? null,
  });

  // Create the result promise
  const resultPromise = new Promise<LLMGenerationResult>((resolve, reject) => {
    resolveResult = resolve;
    rejectResult = reject;
  });

  // Create async generator for tokens
  async function* tokenGenerator(): AsyncGenerator<string> {
    const tokenQueue: string[] = [];
    let resolver: ((value: IteratorResult<string>) => void) | null = null;
    let done = false;
    let error: Error | null = null;

    // Start streaming
    native.generateStream(
      prompt,
      optionsJson,
      (token: string, isComplete: boolean) => {
        if (cancelled) return;

        if (!isComplete && token) {
          // Track first token time
          if (firstTokenTime === null) {
            firstTokenTime = Date.now();
          }

          fullText += token;
          tokenCount++;

          if (resolver) {
            resolver({ value: token, done: false });
            resolver = null;
          } else {
            tokenQueue.push(token);
          }
        }

        if (isComplete) {
          done = true;

          // Build final result
          const endTime = Date.now();
          const latencyMs = endTime - startTime;
          const timeToFirstTokenMs = firstTokenTime ? firstTokenTime - startTime : undefined;
          const tokensPerSecond = latencyMs > 0 ? (tokenCount / latencyMs) * 1000 : 0;

          const finalResult: LLMGenerationResult = {
            text: fullText,
            thinkingContent: undefined,
            inputTokens: Math.ceil(prompt.length / 4),
            tokensUsed: tokenCount,
            modelUsed: 'unknown',
            latencyMs,
            framework: 'unknown', // Backend-agnostic
            tokensPerSecond,
            timeToFirstTokenMs,
            thinkingTokens: 0,
            responseTokens: tokenCount,
          };

          if (resolveResult) {
            resolveResult(finalResult);
          }

          if (resolver) {
            resolver({ value: undefined as unknown as string, done: true });
            resolver = null;
          }

          EventBus.publish('Generation', { type: 'completed' });
        }
      }
    ).catch((err: Error) => {
      error = err;
      done = true;
      if (rejectResult) {
        rejectResult(err);
      }
      if (resolver) {
        resolver({ value: undefined as unknown as string, done: true });
      }
      EventBus.publish('Generation', { type: 'failed', error: err.message });
    });

    // Yield tokens
    while (!done || tokenQueue.length > 0) {
      if (tokenQueue.length > 0) {
        yield tokenQueue.shift()!;
      } else if (!done) {
        const result = await new Promise<IteratorResult<string>>((resolve) => {
          resolver = resolve;
        });
        if (result.done) break;
        yield result.value;
      }
    }

    if (error) {
      throw error;
    }
  }

  // Cancel function
  const cancel = (): void => {
    cancelled = true;
    cancelGeneration();
  };

  return {
    stream: tokenGenerator(),
    result: resultPromise,
    cancel,
  };
}

/**
 * Cancel ongoing text generation
 */
export function cancelGeneration(): void {
  if (!isNativeModuleAvailable()) {
    return;
  }
  const native = requireNativeModule();
  native.cancelGeneration();
}
